pageserver/src/tenant/timeline.rs

```diff
@@ -614,6 +614,31 @@ impl Timeline {
     pub async fn compact(self: &Arc<Self>, ctx: &RequestContext) -> anyhow::Result<()> {
         const ROUNDS: usize = 2;
 
+        static CONCURRENT_COMPACTIONS: once_cell::sync::Lazy<tokio::sync::Semaphore> =
+            once_cell::sync::Lazy::new(|| {
+                tokio::sync::Semaphore::new(usize::max(
+                    1,
+                    // while a lot of the work is done on spawn_blocking, we still do
+                    // repartitioning in the async context. this should give leave us some workers
+                    // unblocked to be blocked on other work, hopefully easing any outside visible
+                    // effects of restarts.
+                    (*task_mgr::RUNTIME_WORKER_THREADS * 3)
+                        .checked_div(4)
```

Comment:
It will be nice to explain why concurrent compaction is limited by 3/4 of workers (not 1/2 or 1/3 fir example)

Reply:
I don't have a good explanation, this is all Harrison-Stetson (guess), but trying to elaborate:

Do you think half or 1/3 would be better? I was thinking there'd be slight increase in perf due and some room to do layer flushes (tenants created right after restart) so I cannot really see. Right now we've been blasting all 8 and unlimited number of L0 => L1 due to `spawn_blocking` but I think for overall create_tenant => create_timeline BACKGROUND_RUNTIME waiting the problem is repartitioning and create_image_layers.

Unsure if I've created issues on this.

Pros for 6/8: I think the perf drop from 8 -> 4 or 3 would be noticeable, but at the same time we don't really track ovrall compaction time. Unsure if we should. IO time seconds metric should not be affected because these should not be in that. We could add an IO time seconds for repartitioning.

Reply:
[0199cb8](https://github.com/neondatabase/neon/pull/4777/commits/0199cb85e60d2d07cc1e1839ea03a6ad96a93c99)